src/computation.ts

```diff
@@ -10,57 +11,58 @@ import { Env, Capabilities, chainEnv, pureEnv, resumeNow, Resume, Use, Embed, re
 // Pure code simply executes between the yields.
 export interface Computation<Y, R, N> {
   'fx-ts/Computation': never
-  [Symbol.iterator](): Iterator<Y, R, N>
+  [Symbol.iterator](): Generator<Y, R, N>
 }
 
 export type Yield<C> = C extends Computation<infer Y, any, any> ? Y : never
 export type Return<C> = C extends Computation<any, infer R, any> ? R : never
 export type Next<C> = C extends Computation<any, any, infer N> ? N : never
 
 type Result<C> = {
-  [K in keyof C]: C[K] extends (...a: readonly any[]) => Resume<infer R> ? R : never
+  [K in keyof C]: C[K] extends (...a: readonly any[]) => Resume<any, infer A> ? A : never
 }[keyof C]
 
 // Create a Computation from an Env-yielding generator
 // allows the computation to be started more than once by calling the
 // generator function each time its iterator is requested
 export const co = <A extends readonly any[], Y, R, N>(f: (...args: A) => Generator<Y, R, never>): ((...args: A) => Computation<Y, R, N>) =>
   (...args: A): Computation<Y, R, N> => ({
-    [Symbol.iterator](): Iterator<Y, R, N> { return f(...args) }
+    [Symbol.iterator](): Generator<Y, R, N> { return f(...args) }
   }) as Computation<Y, R, N>
 
 // Create a Computation from an Env
-export const op = <C, A = Result<C>>(env: Env<C, A>): Computation<Env<C, A>, A, A> => ({
-  *[Symbol.iterator](): Iterator<Env<C, A>, A, A> { return yield env }
-}) as Computation<Env<C, A>, A, A>
+export const op = <C, A = Result<C>, R = never>(env: Env<C, R, A>): Computation<Env<C, R, A>, A, A> => ({
+  *[Symbol.iterator](): Generator<Env<C, R, A>, A, A> { return yield env }
+}) as Computation<Env<C, R, A>, A, A>
 
-export const runComputation = <Y extends Env<any, N>, R, N>(g: Computation<Y, R, N>): Env<Capabilities<Y>, R> =>
-  c => {
+export const runComputation = <Y extends Env<any, R, N>, R, N>(g: Computation<Y, R, N>, c: Capabilities<Y>): Resume<R, R> =>
+  resume(k => {
     const i = g[Symbol.iterator]()
-    return stepComputation(i, i.next(), c)
-  }
+    return stepComputation(i, i.next(), c, k)
+  })
 
-const stepComputation = <Y extends Env<any, N>, R, N>(i: Iterator<Y, R, N>, ir: IteratorResult<Y, R>, c: Capabilities<Y>): Resume<R> => {
+const stepComputation = <Y extends Env<any, R, N>, R, N>(i: Generator<Y, R, N>, ir: IteratorResult<Y, R>, c: Capabilities<Y>, k: (s: Step<R, R>) => Cancel): Cancel => {
   while (true) {
-    if (ir.done) return resumeNow(ir.value)
+    if (ir.done) return k({ done: false, value: ir.value })
```

Comment:
Should this be done=true?

Reply:
Confusingly, no, at least not right now.  Right now, `done` indicates that the current computation should return immediately, much like forcing a generator to return by calling `return` on it.  So, returning `done: true` here will propagate to the enclosing computation, causing it to return, and so on, all the way to outermost computation.  You can see that `done: true` is interpreted to `generator.return` below on line 50.

That upward propagation allows handling "early return" effects at whatever higher level in the computation stack the user chooses, e.g. by applying attempt() at whatever spot they need, much like exceptions and `catch`.

The name `done` is probably extra-confusing here, since it's overloaded with iterator/generator terminology.